lib/pure/concurrency/cpuinfo.nim

```diff
@@ -101,3 +102,207 @@ proc countProcessors*(): int {.rtl, extern: "ncpi$1".} =
   else:
     result = sysconf(SC_NPROCESSORS_ONLN)
   if result <= 0: result = 0
+
+# TODO(awr1): get cache line size
+# TODO(awr1): get processor name
+# TODO(awr1): ARM processor exts
+# TODO(awr1): on ARM, probe /proc/cpuinfo, but what about windows on arm?
+
+const
+  onX86 = defined(i386) or defined(amd64)
+  onArm = defined(arm) or defined(arm64)
+
+when onX86:
+  proc cpuidX86(leaf: int32): tuple[eax, ebx, ecx, edx: int32] =
+    when defined(vcc):
+      # limited inline asm support in vcc, so intrinsics, here we go:
+      proc cpuidVcc(cpuInfo: ptr int32; functionID: int32)
+        {.cdecl, importc: "__cpuid", header: "intrin.h".}
+      cpuidVcc(addr result.eax, leaf)
+    else:
+      var (eaxr, ebxr, ecxr, edxr) = (0'i32, 0'i32, 0'i32, 0'i32)
+      # zero ecx first!
+      asm """
+        xorl %%ecx, %%ecx
+        cpuid
+        :"=a"(`eaxr`), "=b"(`ebxr`), "=c"(`ecxr`), "=d"(`edxr`)
+        :"a"(`leaf`) """
+      (eaxr, ebxr, ecxr, edxr)
+
+when onX86 or defined(nimdoc):
+  type
+    X86Feature* {.pure.} = enum
+      HypervisorPresence, SimultaneousMultithreading, Vmx, X87fpu, Mmx, MmxExt,
+      F3DNow, F3DNowExt, F3DNowPf, Sse, Sse2, Sse3, Ssse3, Sse4a, Sse41, Sse42,
+      Avx, Avx2, Avx512f, Avx512dq, Avx512ifma, Avx512pf, Avx512er, Avx512cd,
+      Avx512bw, Avx512vl, Avx512vbmi, Avx512vbmi2, Avx512vpopcntdq, Avx512vnni,
+      Avx512vnniw4, Avx512fmaps4, Avx512bitalg, Rdrand, Rdseed, MovBigEndian,
+      Popcnt, Fma3, Fma4, Cas8B, Cas16B, Abm, Bmi1, Bmi2, TsxHle, TsxRtm, Adx,
+      Sgx, Gfni, Aes, Vaes, Vpclmulqdq, Pclmulqdq, NxBit, Float16c, Ssbd,
+      SpecCtrl, Stibp, Sha, Clflush, ClflushOpt, Clwriteback, PrefetchWT1, Mpx
+
+when onArm or defined(nimdoc):
+  type
+    ArmFeature* {.pure.} = enum
+      Neon, Vfp
+
+type CpuFeature* = (when onX86: X86Feature elif onArm: ArmFeature)
```

Comment:
This shouldn't be public, we should have a "CPU" object with proc defined for each feature to check to avoid backward compatibility issues when a new feature to detect is introduced/supported

Reply:
Is an object really necessary or could we just calculate every desired feature of these at program startup (if cpuinfo is imported?) and cache the result globally?

Reply:
I meant a private object, not exposed to the public, but it's not necessary for it to be an object either.

Procs could also alternatively use the following scheme:

```Nim
proc cpuHasAVX(): bool =
  let avxCache {.global.} = determineAVX()
  return avxCache
```

And it will compute that only once. I'm not sure what happens if it's called in a parallel loop though so a simple global is probably better

```Nim
let avxCache = determineAVX()

proc cpuHasAVX(): bool =
  return avxCache
```

This way it's called only on module import.

Or alternatively

```Nim
type CpuFeatures = object
  hasSSE3: bool
  hasSSE4_1: bool
  hasAVX: bool
  hasAVX2: bool
  hasAVX512: bool

# initialized on module import
let cacheCpuFeatures = determineCpuFeatures()

proc hasAVX*(): bool =
  return cacheCpuFeatures.hasAVX()
```

THis is how it's done in Facebook's cpuinfo: https://github.com/pytorch/cpuinfo/blob/40c5f3695b053e5c3d642d9bc34113f3baa71ef2/include/cpuinfo.h#L579-L682.
Nim does have sets but the iteration through the native set is costly for such big sets as I said in the review.